src/main/java/com/amazon/dlic/auth/ldap/util/LdapHelper.java

```diff
@@ -35,36 +35,45 @@
 
 import org.opensearch.SpecialPermission;
 
+
 public class LdapHelper {
 
     private static SearchFilter ALL = new SearchFilter("(objectClass=*)");
+
     @SuppressWarnings("removal")
     public static List<LdapEntry> search(final Connection conn, final String unescapedDn, SearchFilter filter,
-            final SearchScope searchScope, final String[] returnAttributes) throws LdapException {
+            final SearchScope searchScope, final String[] returnAttributes, boolean shouldFollowReferrals) throws LdapException {
 
         final SecurityManager sm = System.getSecurityManager();
 
         if (sm != null) {
             sm.checkPermission(new SpecialPermission());
         }
-
         try {
             final String baseDn = escapeDn(unescapedDn);
             return AccessController.doPrivileged(new PrivilegedExceptionAction<List<LdapEntry>>() {
                 @Override
                 public List<LdapEntry> run() throws Exception {
                     final List<LdapEntry> entries = new ArrayList<>();
                     final SearchRequest request = new SearchRequest(baseDn, filter);
-                    request.setReferralHandler(new SearchReferralHandler());
+
                     request.setSearchScope(searchScope);
                     request.setDerefAliases(DerefAliases.ALWAYS);
                     request.setReturnAttributes(returnAttributes);
                     final SearchOperation search = new SearchOperation(conn);
-                    // referrals will be followed to build the response
-                    final Response<SearchResult> r = search.execute(request);
-                    final org.ldaptive.SearchResult result = r.getResult();
-                    entries.addAll(result.getEntries());
-                    return entries;
+
+                    if (shouldFollowReferrals) {
+                        request.setReferralHandler(new SearchReferralHandler());
+                        final Response<SearchResult> r = search.execute(request);
+                        // referrals will be followed to build the response
+                        final org.ldaptive.SearchResult result = r.getResult();
+                        entries.addAll(result.getEntries());
+                        return entries;
+                    } 
+                    else{
+                        entries.add(new LdapEntry("cn=couldNotFollowReferral"));
```

Comment:
I don't think adding entries like this will be helpful to callers of the API.  I think the test can be updated to check for the lack of a referral retrieved value, what do you think?